### PR TITLE
Add H-Bridge switch component

### PR DESCRIPTION
**Pull request by dwmw2**
# What does this implement/fix?

Add an H-Bridge switch component for driving the relay in devices like https://devices.esphome.io/devices/Tongou-TO-Q-SY1-JWT

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/esphome/discussions/4087

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/4233
## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [x] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

switch:
  - platform: hbridge
    on_pin: P24
    off_pin: P26
    id: relay
    # On by default should be fine as this is not a safety device
    restore_mode: RESTORE_DEFAULT_ON
    pulse_length: 100ms
    wait_time: 50ms
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by dwmw2** [comment]
There is some debate about whether "H-Bridge" is the right name for this. It certainly matches OpenBeken, and the feature requests such as https://github.com/esphome/feature-requests/issues/1690 and https://github.com/esphome/esphome/discussions/4087 so it looks like it would be well-named for users to actually find it?

Not sure it's strictly technically accurate though, and maybe it would be better as `platform: bistable ` or something like that?

**Comment by raidnet-ms** [comment]
Adding a delay time for on and off signal would be great.

About the name well If i search anything related with the reverse of polarity I look for h-bridge keyword.

Bistable solenoid is only a usecase but there could be more (open and close motorized valves for istance)

**Comment by dwmw2** [comment]
> Adding a delay time for on and off signal would be great.

Can you help me understand what you need that `pulse_length` and `wait_time` don't already cover? (I have now added them to the example YAML above.)

Docs at 
https://deploy-preview-4233--esphome.netlify.app/components/switch/hbridge.html

**Comment by raidnet-ms** [comment]
> > Adding a delay time for on and off signal would be great.
> 
> Can you help me understand what you need that `pulse_length` and `wait_time` don't already cover? (I have now added them to the example YAML above.)
> 
> Docs at 
> https://deploy-preview-4233--esphome.netlify.app/components/switch/hbridge.html

Sure they are perfect. For some reason I didn't see those values before!!

File: esphome/components/hbridge/switch/hbridge_switch.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  LOG_PIN("  On Pin: ", this->on_pin_);
+  LOG_PIN("  Off Pin: ", this->off_pin_);
+  ESP_LOGCONFIG(TAG, "  Pulse length: %" PRId32 " ms", this->pulse_length_);
+  if (this->wait_time_)
+    ESP_LOGCONFIG(TAG, "  Wait time %" PRId32 " ms", this->wait_time_);
+}
+
+void HBridgeSwitch::write_state(bool state) {
+  this->desired_state_ = state;
+  if (!timer_running_)
```
Comment (tomaszduda23) [inline esphome/components/hbridge/switch/hbridge_switch.cpp]: this->timer_running_

**Comment by sorrypqa** [comment]
thank you! This component is very practical.

**Comment by sorrypqa** [comment]
Can we add zero crossing control function for AC power and specify a zero crossing detection IO and delay time. When zero crossing is detected, there is a delay of XXmS, and the relay performs opening or closing actions, which reduces sparks, minimizes interference, and makes the relay more durable.

**Comment by sorrypqa** [comment]
The GS1102 chip can be used for zero crossing detection, providing low-level trigger pulses

**Comment by dwmw2** [comment]
> Can we add zero crossing control function for AC power and specify a zero crossing detection IO and delay time. When zero crossing is detected, there is a delay of XXmS, and the relay performs opening or closing actions, which reduces sparks, minimizes interference, and makes the relay more durable.

I'd quite like to get the basic functionality merged; it's been pending for a long time now. But I'm happy to explore this idea too. How would it work? 

One option is to use the GS1102 as an input and then trigger the H-Bridge switch when zero is detected, but the delay of doing all that through software probably makes it unworkable. So... instead of GPIO output pins running directly to the open/close pins of H-Bridge switch, you'd route each through an AND gate, so they're only active at the zero point? What exactly would that mean for this component? You'd want the zero detection also to be an input, and the pulse would last until that input is triggered instead of a fixed length of time?

**Comment by sorrypqa** [comment]
> GS1102

The function I hope for is:
The H-bridge component specifies the zero crossing detection IO and enables the zero crossing trigger function. Whether the H-bridge switch is turned on or off, pin_forward or pin_deverse is in a waiting state until the zero crossing detection IO is triggered by the low-level zero crossing signal of GS1102. After a custom delay time, pin_forward or pin_deverse outputs the relay drive signal

**Comment by dwmw2** [comment]
Do we need to build that into the H-Bridge switch component? Wouldn't that be a useful thing for *any* switch, using a separate zero-detection? I'm imagining a generic `switch-when-zero` switch type, which takes a *slave* switch type and a trigger GPIO input, and sets the slave to match its own on/off state whenever the zero trigger is asserted?

**Comment by sorrypqa** [comment]
That is to say, the H-bridge component has two options:
1. Directly triggered, like the current function, whether the H-bridge switch is turned on or off, pin_fforwards or pin_deverse will directly output the driving signal
2. Zero crossing trigger, specifying the zero crossing detection IO and setting a delay time. Whether the H-bridge switch is turned on or off, pin_forward or pin_deverse is in a waiting state until the zero crossing detection IO is triggered by the low-level zero crossing signal of GS1102. After a custom delay time, pin_forward or pin_deverse outputs the relay drive signal

**Comment by sorrypqa** [comment]
Because magnetic holding relays are energy-saving and do not generate heat, they are now quite popular, and this H-bridge switch component is perfect for showcasing its capabilities.

**Comment by dwmw2** [comment]
Sure, but the zero-crossing thing is surely orthogonal to that? What if we build a *generic* implementation of that?

```
 switch:
   - platform: trigger
     name: "Zero-triggered switch"
     trigger_pin: GPIO10 # The zero detection pin
     subordinate_switch_id: the_real_one # Either an hbridge switch or any other type
     trigger_delay: 10ms
   - platform: whatever
     id: the_real_one
     …
```
I'd knock up an implementation of that to play with, but contributing to ESPHome isn't sufficiently feeding my dopamine addition because things are taking so long to get merged :)

**Comment by sorrypqa** [comment]
> Sure, but the zero-crossing thing is surely orthogonal to that? What if we build a _generic_ implementation of that?
> 
> ```
>  switch:
>    - platform: trigger
>      name: "Zero-triggered switch"
>      trigger_pin: GPIO10 # The zero detection pin
>      subordinate_switch_id: the_real_one # Either an hbridge switch or any other type
>      trigger_delay: 10ms
>    - platform: whatever
>      id: the_real_one
>      …
> ```
> 
> I'd knock up an implementation of that to play with, but contributing to ESPHome isn't sufficiently feeding my dopamine addition because things are taking so long to get merged :)

It looks pretty good and I can't wait to try it out. I wonder if this series control method can accurately and quickly respond,Because the zero crossing point of alternating current changes too quickly

**Comment by kbx81** [comment]
I think this looks good -- @dwmw2 any chance you'd be willing to consolidate the tests into a single `common.yaml` file? Have a look at the [bme280](https://github.com/esphome/esphome/tree/dev/tests/components/bme280_i2c) for a hint.

**Comment by dwmw2** [comment]
ISTR I had deliberately made the tests different, to exercise different combinations of options. Will take a look...

**Comment by kbx81** [comment]
> ISTR I had deliberately made the tests different, to exercise different combinations of options. Will take a look...

You can still accomplish that -- use [extend and/or remove](https://esphome.io/components/packages#config-packages-extend) if necessary.
Thanks for working on this! 🍻

**Comment by dwmw2** [comment]
Attempted the common template in https://github.com/esphome/esphome/pull/7421/commits/49c8ab5a7d575d34b67bc652811169b3f798193b

**Comment by dwmw2** [comment]
> > ISTR I had deliberately made the tests different, to exercise different combinations of options. Will take a look...
> 
> You can still accomplish that -- use [extend and/or remove](https://esphome.io/components/packages#config-packages-extend) if necessary. Thanks for working on this! 🍻

@kbx81 I had to do the `!include` differently to the bme280 example you gave, because it wasn't managing to extend the configuration. But it should now be working with that latest 'Make tests actually pass...' commit.

File: esphome/components/hbridge/switch/hbridge_switch.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ protected:
+  void write_state(bool state) override;
+  void timer_fn_();
+
+  bool timer_running_{false};
+  bool desired_state_;
+  RelayState relay_state_{RELAY_STATE_UNKNOWN};
+  GPIOPin *on_pin_;
+  GPIOPin *off_pin_;
+  uint32_t pulse_length_;
```
Comment (kbx81) [inline esphome/components/hbridge/switch/hbridge_switch.h]:
```suggestion
  bool desired_state_{false};
  RelayState relay_state_{RELAY_STATE_UNKNOWN};
  GPIOPin *on_pin_{nullptr};
  GPIOPin *off_pin_{nullptr};
  uint32_t pulse_length_{0};
```

Comment (kbx81) [reply esphome/components/hbridge/switch/hbridge_switch.h]: I was just doing a bunch of clean-up like this elsewhere...

**Review by kbx81 — APPROVED** [review]
LGTM -- thanks again! 🍻

**Comment by dwmw2** [comment]
> I wonder if this series control method can accurately and quickly respond,Because the zero crossing point of alternating current changes too quickly

Yes, as I said [above](https://github.com/esphome/esphome/pull/7421#issuecomment-2478511111) you probably want hardware assistance. But even with hardware assistance for the trigger I'm not sure a latching H-Bridge switch will work well for that use case. These things typically have a pulse length of 100ms — meaning that a 50Hz signal will have completed five full cycles (crossing zero ten times) in the time it takes to know that the switch has completed its movement.

**Pull request merged**